index.js

```diff
@@ -6,63 +6,54 @@ var glob = require('glob');
 var arrify = require('arrify');
 var pify = require('pify');
 
-function sortPatterns(patterns) {
-	patterns = arrify(patterns);
+var globP = pify(glob, Promise).bind(glob);
 
-	var positives = [];
-	var negatives = [];
+function isNegative(pattern) {
+	return pattern[0] === '!';
+}
 
-	patterns.forEach(function (pattern, index) {
-		var isNegative = pattern[0] === '!';
-		(isNegative ? negatives : positives).push({
-			index: index,
-			pattern: isNegative ? pattern.slice(1) : pattern
-		});
-	});
+function generateGlobTasks(patterns, opts) {
+	var globTasks = [];
 
-	return {
-		positives: positives,
-		negatives: negatives
-	};
-}
+	patterns = arrify(patterns);
+	opts = objectAssign({ignore: []}, opts);
 
-function setIgnore(opts, negatives, positiveIndex) {
-	opts = objectAssign({}, opts);
+	patterns.forEach(function (pattern, i) {
+		if (isNegative(pattern)) {
+			return;
+		}
 
-	var negativePatterns = negatives.filter(function (negative) {
-		return negative.index > positiveIndex;
-	}).map(function (negative) {
-		return negative.pattern;
+		var ignore = patterns.slice(i).filter(isNegative).map(function (pattern) {
```

Comment:
I agree, the file system is the real bottleneck. I prefer my version, because it is easier to grasp how negative patterns have an effect, that is clearly opinionated though.

If you think it is worth, I suggest instead of the
```js
var negativePatterns = negatives.filter(function (negative) {
	return negative.index > positiveIndex;
}).map(function (negative) {
	return negative.pattern;
});
```
logic, some caching strategy:
```js
var negativePatterns = negatives.slice(negativesOf[i]);
```

But I'm not sure about significant overall speed advantages.